core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java

```diff
@@ -0,0 +1,268 @@
+        }
+        return Optional.empty();
+    }
+
+    private String getHostForComponentInService(DatalakeResources datalakeResources, String service, String serviceComponent) {
+        if (datalakeResources.getServiceDescriptorMap().containsKey(service)) {
+            ServiceDescriptor serviceDescriptor = datalakeResources.getServiceDescriptorMap().get(service);
+            Map<String, Object> componentHostMap = serviceDescriptor.getComponentsHosts().getMap();
+            return String.valueOf(componentHostMap.get(serviceComponent));
+        } else {
+            return null;
+        }
+    }
+
+    private String getRangerAdminHost(DatalakeResources datalakeResources) {
+        return getHostForComponentInService(datalakeResources, RANGER, RANGER_ADMIN);
+    }
+
+    private String getRangerPort(Map<String, ServiceDescriptor> serviceDescriptorMap, String defaultPort) {
+        if (serviceDescriptorMap.containsKey(RANGER)) {
+            ServiceDescriptor serviceDescriptor = serviceDescriptorMap.get(RANGER);
+            Map<String, Object> params = serviceDescriptor.getBlueprintParams().getMap();
+            return String.valueOf(params.getOrDefault(RANGER_PORT, defaultPort));
+        }
+        return defaultPort;
+    }
+
+    private String getServiceRelatedSecret(Map<String, ServiceDescriptor> serviceDescriptorMap, String service, String secretKey) {
+        if (serviceDescriptorMap.containsKey(service)) {
+            ServiceDescriptor serviceDescriptor = serviceDescriptorMap.get(service);
+            Map<String, Object> secretMap = serviceDescriptor.getBlueprintSecretParams().getMap();
+            if (secretMap.containsKey(secretKey)) {
+                return String.valueOf(secretMap.get(secretKey));
+            }
+        }
+        return null;
+    }
+
+    private String getServiceRelatedSecret(DatalakeResources datalakeResources, String service, String secretKey) {
+        return getServiceRelatedSecret(datalakeResources.getServiceDescriptorMap(), service, secretKey);
+    }
+
+    private String getRangerAdminPassword(DatalakeResources datalakeResources) {
+        return getServiceRelatedSecret(datalakeResources, RANGER, RANGER_PASSWORD);
+    }
+
+    private DatalakeResources getDatalakeResourceId(StackV4Request source, Workspace workspace) {
+        try {
+            if (source.getSharedService() != null && isNotBlank(source.getSharedService().getDatalakeName())) {
+                return datalakeResourcesService.getByNameForWorkspace(source.getSharedService().getDatalakeName(), workspace);
+            }
+        } catch (NotFoundException nfe) {
+            LOGGER.debug("No datalake resource found for data lake: {}", source.getSharedService().getDatalakeName());
+        }
+        return null;
+    }
+
+    private SharedServiceConfigsView createSharedServiceConfigs(Blueprint blueprint, String ambariPassword,
+            Optional<DatalakeResources> datalakeResources) {
```

Comment:
in general, passing an Optional as a method argument is not a very welcomed way since it can be tricky because one has to check whether the actual Optional is not null or not, and only after that can be checked with isPresent.

Reply:
it is a private function, it won't be a problem